sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -0,0 +1,530 @@
+    public boolean requiresDeduping() {
+      return spec.useCorrelationId();
+    }
+  }
+
+  private static class RabbitMQCheckpointMark
+      implements UnboundedSource.CheckpointMark, Serializable {
+    transient Channel channel;
+    Instant oldestTimestamp;
+    final List<Long> sessionIds = new ArrayList<>();
+
+    @Override
+    public void finalizeCheckpoint() throws IOException {
+      for (Long sessionId : sessionIds) {
+        channel.basicAck(sessionId, false);
+      }
+      channel.txCommit();
+      oldestTimestamp = Instant.now();
+      sessionIds.clear();
+    }
+  }
+
+  private static class UnboundedRabbitMqReader
+      extends UnboundedSource.UnboundedReader<RabbitMqMessage> {
+    private final RabbitMQSource source;
+
+    private RabbitMqMessage current;
+    private byte[] currentRecordId;
+    private Connection connection;
+    private Channel channel;
+    private QueueingConsumer consumer;
+    private Instant currentTimestamp;
+    private final RabbitMQCheckpointMark checkpointMark;
+
+    UnboundedRabbitMqReader(RabbitMQSource source, RabbitMQCheckpointMark checkpointMark) {
+      this.source = source;
+      this.current = null;
+      this.checkpointMark = checkpointMark != null ? checkpointMark : new RabbitMQCheckpointMark();
+    }
+
+    @Override
+    public Instant getWatermark() {
+      return checkpointMark.oldestTimestamp;
+    }
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return checkpointMark;
+    }
+
+    @Override
+    public RabbitMQSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public byte[] getCurrentRecordId() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      if (currentRecordId != null) {
+        return currentRecordId;
+      } else {
+        return "".getBytes(StandardCharsets.UTF_8);
+      }
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (currentTimestamp == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+    @Override
+    public RabbitMqMessage getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      try {
+        ConnectionFactory connectionFactory = createConnectionFactory(source.spec.uri());
+        connection = connectionFactory.newConnection();
+        channel = connection.createChannel();
+        if (channel == null) {
+          throw new IOException("No RabbitMQ channel available");
+        }
+
+        String queueName = source.spec.queue();
+        if (source.spec.queueDeclare()) {
+          // declare the queue (if not done by another application)
+          // channel.queueDeclare(queueName, durable, exclusive, autoDelete, arguments);
+          channel.queueDeclare(queueName, false, false, false, null);
```

Comment:
Per below it seems that queueName can be null?

Reply:
Still relevant

Reply:
just assert the queue is not null in withQueueDeclare